coqlspclient/coq_structs.py

```diff
@@ -90,12 +90,41 @@ def __hash__(self) -> int:
         return hash(self.text)
 
 
-class ProofStep(Step):
-    def __init__(self, step: Step, goals: GoalAnswer, context: List[Term]):
-        super().__init__(step.text, step.ast)
-        self.goals = goals
+class ProofStep:
+    def __init__(
+        self,
+        step: Step,
+        goals: Union[GoalAnswer, Callable[[Position], GoalAnswer]],
+        context: List[Term],
+    ):
+        self.step = step
+        self._goals = goals
         self.context = context
 
+    @property
+    def goals(self) -> GoalAnswer:
+        if callable(self._goals):
+            goals = self._goals(self.ast.range.start)
+            self._goals = goals
+            return self._goals
```

Comment:
Return can be removed and make it a one liner
```suggestion
            self._goals = self._goals(self.ast.range.start)
```